Add clarifiication on uncle headers (#158)
ethereum/portal-network-specs: history-network.md

encoded_uncles          = rlp.encode(list_of_uncle_headers)
```

Note 1: The type-specific encoding might be different in future transaction types, but this encoding
works for all current transaction types.

> Note 2: The `list_of_uncle_headers` refers to the array of uncle headers [defined in the devp2p spec](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#block-encoding-and-validity).  

Reviewer: Using devp2p as the reference is not wrong, but also not helpful, maybe? It's really the yellow paper spec that defines it. That format is consensus-relevant, because the uncles hash found in the block header is the hash of `rlp.encode(list_of_uncle_headers)`. That's the primary reason for the choice. I would have a hard time using "devp2p did it this way" as a primary justification for any of our design choices :P

Reviewer: The attempt to describe the rationale is here: https://github.com/ethereum/portal-network-specs/blob/e807eb09d2859016e25b976f082735d3aceceb8e/history-network.md#encoding-content-values-for-validation

Author: Ah, gotcha, sorry.  I remember that in your PR and I'm good with it.  I just wanted to explain what `list_of_uncle_headers` actually was since it's not clearly defined that it's an array of byte arrays where each byte array is the raw bytes representation of the uncle header.

Author: I revised my reference to point to the yellow paper and then further define in hopefully clear language how the uncles should be represented.

Reviewer: Cool looks good, thanks for adding!

Note 1: The type-specific encoding might be different in future transaction types, but this encoding
works for all current transaction types.

> Note 2: The `list_of_uncle_headers` refers to the list of uncle headers [defined in the Ethereum yellow paper](https://github.com/ethereum/yellowpaper/blob/30782852fef61f61a247dafbdad814ae7e00fb39/Paper.tex#L399) and is represented as an array of byte arrays where each byte array is the raw bytes representing an uncle header.  

Reviewer 2: I think you rather need to say here that it is an RLP encoded list of block headers (from the uncle blocks) that results in a byte string/array. 

E.g.:
```suggestion
Note 2: The `list_of_uncle_headers` refers to the list of uncle headers [defined in the Ethereum yellow paper](https://github.com/ethereum/yellowpaper/blob/30782852fef61f61a247dafbdad814ae7e00fb39/Paper.tex#L399) and is represented RLP encoded list of block headers, where each block header is the header from an uncle block.
```

Also, I think a bit more relevant yellow paper link is this line: https://github.com/ethereum/yellowpaper/blob/30782852fef61f61a247dafbdad814ae7e00fb39/Paper.tex#L414 ?

Author: I'd be hesitant to say it's an RLP encoded list since that's kind of what the larger `content: rlp(list_of_uncle_headers)` already says and I'm talking specifically about what's inside the parentheses.  I wouldn't want implementers to think we mean our actual code should be `rlp.encode(rlp.encode(list_of_uncle_headers))`.  Otherwise, I'm good with the above.

Reviewer 2: Right, forgot about that. Then I'd suggest dropping everything from the `and` onwards. I mostly wanted to point out that the byte array wording is not really correct.

Author: I went with @carver's suggestion.  Any last comments before we merge this?

> Note 2: The `list_of_uncle_headers` refers to the list of uncle headers [defined in the Ethereum yellow paper](https://github.com/ethereum/yellowpaper/blob/30782852fef61f61a247dafbdad814ae7e00fb39/Paper.tex#L399) and is represented as an array of byte arrays where each byte array is the raw bytes representing an uncle header.  

Reviewer: Yeah, the way @kdeme describes it pretty natural to me -- with the clarification that "an RLP encoded list of block headers (from the uncle blocks)" is the definition of `encoded_uncles`, not `list_of_uncle_headers`.

I'd love to find a version that we all find helpful and informative. Here's another take on the text:
```suggestion
Note 2: The `list_of_uncle_headers` refers to the list of uncle headers [defined in the Ethereum yellow paper](https://github.com/ethereum/yellowpaper/blob/30782852fef61f61a247dafbdad814ae7e00fb39/Paper.tex#L414). As usual, when encoding to RLP, each header must first be formatted as a list where each field is represented as an integer or byte-string. So `list_of_uncle_headers` is a list of these lists, before rlp-encoding.  
```
Part of me thinks we can drop the whole explanation. I wouldn't feel a need to explain `rlp.encode(header)` in a spec, for example. That also requires you to convert each of the fields of the header into an `int` or `bytes`, too.

I can understand the desire to help folks remember how RLP works with lists of objects, which is a little less common, though. Then, it's probably important to also note that some of the fields are represented as integers instead of bytes.

The "As usual" is a nice cue for people who recognize this as standard practice for rlp-encoding lists. Without that, it reads like we're trying to tell folks that some abnormal encoding is happening, instead.

Author: Any of these outcomes is fine with me ultimately.  I just know it took me a minute to get what was going on and I didn't recognize the `list_of_uncle_headers` reference back to the yellow paper (though maybe that's more explainable by my not understanding core concepts well enough more than anything).
